plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java

```diff
@@ -273,79 +279,185 @@ private static boolean isDirectory(URI uri)
         }
 
         @Override
         public synchronized long getRetainedSize()
         {
-            return INSTANCE_SIZE + (sliceInput == null ? 0 : sliceInput.getRetainedSize());
+            return INSTANCE_SIZE + bufferRetainedSize;
         }
 
         @Override
-        public synchronized boolean isFinished()
+        public boolean isFinished()
         {
             return closed;
         }
 
         @Override
-        public synchronized void close()
+        public void close()
         {
             if (closed) {
                 return;
             }
             closed = true;
+
+            currentFile = null;
+            sliceInput = null;
+            bufferRetainedSize = 0;
+            inProgressReadFuture.cancel(true);
+            inProgressReadFuture = immediateVoidFuture(); // such that we don't retain reference to the buffer
+        }
+
+        private void fillBuffer()
+        {
+            if (currentFile == null || fileOffset == currentFile.getFileSize()) {
+                currentFile = sourceFiles.poll();
+                if (currentFile == null) {
+                    close();
+                    return;
+                }
+                fileOffset = 0;
+            }
+
+            byte[] buffer = new byte[bufferSize];
+            int bufferFill = 0;
             if (sliceInput != null) {
-                sliceInput.close();
-                sliceInput = null;
+                int length = sliceInput.available();
+                sliceInput.readBytes(buffer, 0, length);
+                bufferFill += length;
+            }
+
+            ImmutableList.Builder<ListenableFuture<Void>> downloadFutures = ImmutableList.builder();
+            while (true) {
+                long fileSize = currentFile.getFileSize();
+                // Make sure Azure Blob Storage read request byte ranges align with block sizes for best performance
+                int readableBlocks = (buffer.length - bufferFill) / blockSize;
+                if (readableBlocks == 0) {
+                    if (buffer.length - bufferFill >= fileSize - fileOffset) {
+                        readableBlocks = 1;
+                    }
+                    else {
+                        break;
+                    }
+                }
+
+                BlockBlobAsyncClient blockBlobAsyncClient = blobServiceAsyncClient
+                        .getBlobContainerAsyncClient(getContainerName(currentFile.getFileUri()))
+                        .getBlobAsyncClient(getPath(currentFile.getFileUri()))
+                        .getBlockBlobAsyncClient();
+                Optional<SecretKey> secretKey = currentFile.getSecretKey();
+                if (secretKey.isPresent()) {
+                    blockBlobAsyncClient = blockBlobAsyncClient.getCustomerProvidedKeyAsyncClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+                }
+                for (int i = 0; i < readableBlocks && fileOffset < fileSize; ++i) {
```

Comment:
Do we need `&& fileOffset < fileSize` here? Isn't how `readableBlocks` is computed enough to not go over file size?

Reply:
We need. readableBlocks is calculated based on the remaining space in the buffer, but it's possible that the remaining file content is even shorter